doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,210 @@
+are at the exact same point. The existence of the slot will prevent any WAL
+segments from being cleaned up while our replica is busy reading the initial
+snapshot.
+
+Once the initial sync is done the transaction is committed to release the
+snapshot and the replica can start streaming the slot normally with no data
+loss.
+
+### Transaction boundaries and timestamps
+
+A key design goal is to maintain the  transaction boundaries from the upstream
+database so that no query on materialize can observe states that never existed
+upstream. At the same time, the data ingested from postgres should be join-able
+with other sources in materialize. This means that all updates must have a
+timestamp that is related to the wall-clock time.
+
+The implementation is greatly simplified by the logical replication protocol
+that guarantees that we'll receive full transactions in commit order. In the
+replication stream each transaction is represented by a BEGIN message, followed
+by one or more row updates, and finally by a COMMIT message.
+
+Each transaction is timestamped with the wall-clock time at the time of the
+BEGIN message. This allows the source implementation to emit rows as it receives
+them and ensure that transaction boundaries are guaranteed.
+
+### Adaptation to multiple tables
+
+If the sources naively applied the mechanism described above for every source
+instance they would very quickly exhaust the available replication slots from
+upstream and also multiply the bandwidth used by several times.
+
+In order to have one independent `SOURCE` object for each upstream table while
+also maintaining a single replication connection between them a new SQL syntax
+will be introduced that allows creating multiple sources at once.
+
+Specifically, a user should be able to do `CREATE SOURCES FROM POSTGRES..` and
+have one `SOURCE` object created for each upstream table. By introducing this
+syntax the SQL planner will have an opportunity to create all the relevant
+sources at once and therefore give them some shared state. The shared state
+could be some form of `Arc<RwLock<Option<T>>>` where the first source instance
+that grabs the write lock will initialize it and afterwards all the source
+instances will keep a read lock around to share the stream.
+
+The concrete syntax for `CREATE SOURCES` can be seen in
+https://github.com/MaterializeInc/materialize/pull/6299.
+
+## Open questions
+
+### Robust resume logic
+
+A strategy for handling connection disconnects is missing for both the initial
+syncing phase and the streaming phase.
+
+The initial syncing phase is the most challenging one but also the least
+important one as this only happens on source startup. In order to handle
+disconnections during this phase the implementation needs to be able to issue a
+brand new COPY command at the same point as the interrupted transfer. Normally,
+a system could just flush its in-flight state and start anew, but since we're
```

Comment:
Would it be unreasonable to drop and re-ship a dataflow instead of issuing retractions?

Reply:
Can this be done safely? What if the data has reached some external sink?